{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "9198e9a5-e26c-4ccb-bc4e-9fac160f96e8",
   "metadata": {},
   "source": [
    "# Annotation of genomes\n",
    "\n",
    "Once we have analysis ready genomes (verified quality and lack of contamination), we can do annotation. The main goal of annotation is to identify the genomic features that are present in the analysis contigs. Usually, these tools use predictive algorithms (recognition of common sequences in the DNA) or alignment against a reference database to identify the different features of a genome. \n",
    "\n",
    "Bakta is a tool that produces a comprehensive annotation composed of Open Reading Frames (sequences that include codons for aminoacids of at least certain length), different types of RNA (ncRNA, tRNA, rRNA), hypothetical proteins, genes and CRISPR arrays. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eeaab644",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "# source PATH to use module function\n",
    "source /cvmfs/soft.computecanada.ca/config/profile/bash.sh"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "ab7b4637",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "Bakta uses a reference database that may be really large (~30GB). A reduced database of approximately 3.0 GB can be employed and will produce less specific results in feature annotation. \n",
    "\n",
    "We already downloaded the database using the command below and it is available in `/tools/reference_data/`\n",
    "```\n",
    "singularity exec -B /scratch,/project,/etc </PATH/TO/BAKTA/IMAGE/> \\\n",
    "    bakta_db download \\\n",
    "    --output </OUTPUT/DIR/> \\\n",
    "    --type light\n",
    "```"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "aba8980b",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "source": [
    "We save the path to the bakta singularity image and reference dataset in the following way to provide it in the main pipeline.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "235e8d53-d8cb-4289-8e41-c9f98d8e89bb",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "BAKTA_IMG=\"/mnt/cidgoh-object-storage/images/bakta_1.7.sif\"\n",
    "export BAKTA_DB=\"/mnt/cidgoh-object-storage/database/bakta/db.tar.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d0309dd4-fa2b-4a4d-af80-145bc45e48a7",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "singularity exec -B /etc \"$BAKTA_IMG\" bakta \\\n",
    "    --db \"$BAKTA_DB\"                                                        `# path to bakta database` \\\n",
    "    /home/jupyter-mdprieto/tutorials/contigs/ERR10479518_contigs.fa         `# file to annotate`\\\n",
    "    --output /home/jupyter-mdprieto/tutorials/annotation/                   `# output directory` \\\n",
    "    --genus Pseudomonas                                                     `# specify genus of isolate` \\\n",
    "    --prefix $(echo \"ERR10479518_contigs.fa\" | grep -Eo \"ERR[0-9]+\")        `# prefix of sample name only for output`"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "92b54117",
   "metadata": {},
   "source": [
    "Now, our analysis results look something like this. The directory structure now includes results from annotation and the tools directory includes a reference database to run the annotation. \n",
    "\n",
    "```\n",
    "/home/jupyter-mdprieto\n",
    "|-- scripts\n",
    "|-- tools\n",
    "|   |-- reference_data\n",
    "|   `-- sing_imgs\n",
    "`-- tutorials\n",
    "    |-- annotation\n",
    "    |-- assembly_checkm\n",
    "    |-- assembly_quast\n",
    "    |-- contigs\n",
    "    |-- raw_reads\n",
    "    |-- results_qc\n",
    "    `-- trimmed_reads\n",
    "```\n",
    "\n",
    "The output of **Bakta** annotation can be found in `tutorials/annotation`. Circular plots can be visualized by Jupyter by clicking them or exported to a local system. They may be partially incomplete as we are working with draft genomes instead of complete chromosome asssemblies or scaffolds. These more refined assemblies require long reads to polish the contigs. \n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "980567bf",
   "metadata": {},
   "source": [
    "# Phylogenetic tree\n",
    "\n",
    "To produce a phylogenetic tree we must compare the genomes using some kind of metric. Here, we will use **Snippy** to identify single nucleotide variants in our genomes and compare how different are these isolates between them. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "e53365b4",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "SNIPPY_IMG=\"/home/jupyter-mdprieto/tools/sing_imgs/snippy_4.6.0.sif\"\n",
    "GUBBINS_IMG=\"/home/jupyter-mdprieto/tools/sing_imgs/gubbins-3.2.1.img\"\n",
    "CONTIGS_DIR=\"/home/jupyter-mdprieto/tutorials/contigs\"\n",
    "REF_GENOME=\"/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna\""
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "dafef90b",
   "metadata": {},
   "source": [
    "Snippy is a relatively fast tool, so we can analyze all the genomes we have available without too much wait. We apply the tool to every available draft genome. **Snippy** can also be used with raw reads and is actually the recommended use."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "ac302678",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 253774 sequences (63420526 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.732 CPU sec, 4.157 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 253774 reads in 29.132 CPU sec, 3.567 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479510\\tSM:ERR10479510 -t 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.843 sec; CPU: 64.386 sec\n",
      "[samclip] Total SAM records 574968, removed 9431, allowed 1610, passed 565537\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 57\n",
      "          no. right trimmed                     : 388\n",
      "          no. left and right trimmed            : 17\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 462\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 462\n",
      "       total no. variants observed              : 27311\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.44s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "The Snippy manual is at http://github.com/tseemann/snippy/blob/master/README.md\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 253688 sequences (63394572 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 34.427 CPU sec, 4.251 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 253688 reads in 31.425 CPU sec, 3.840 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479511\\tSM:ERR10479511 -t 8[samclip] Total SAM records 574979, removed 9845, allowed 1597, passed 565134\n",
      " reference/ref.fa[samclip] Header contained 3 lines\n",
      " fake_reads.fq[samclip] Done.\n",
      "\n",
      "[main] Real time: 9.807 sec; CPU: 67.019 sec\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 58\n",
      "          no. right trimmed                     : 390\n",
      "          no. left and right trimmed            : 18\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 466\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 466\n",
      "       total no. variants observed              : 27283\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.44s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "Have a suggestion? Tell me at http://github.com/tseemann/snippy/issues\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 253030 sequences (63239792 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.697 CPU sec, 4.173 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 253030 reads in 29.121 CPU sec, 3.567 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479512\\tSM:ERR10479512 -t 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.655 sec; CPU: 64.046 sec\n",
      "[samclip] Total SAM records 574186, removed 9514, allowed 1594, passed 564672\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 59\n",
      "          no. right trimmed                     : 386\n",
      "          no. left and right trimmed            : 19\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 464\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 464\n",
      "       total no. variants observed              : 27298\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.43s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "Run 'snp-sites -b -c -o phylo.aln core.full.aln' for IQTree, BEAST, RaxML\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 252942 sequences (63206218 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.411 CPU sec, 4.121 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 252942 reads in 28.998 CPU sec, 3.547 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD:[samclip] Total SAM records 574178, removed 9556, allowed 1657, passed 564622\n",
      " bwa mem[samclip] Header contained 3 lines\n",
      " -Y[samclip] Done.\n",
      " -M -R @RG\\tID:ERR10479513\\tSM:ERR10479513 -t 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.498 sec; CPU: 63.605 sec\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 58\n",
      "          no. right trimmed                     : 387\n",
      "          no. left and right trimmed            : 19\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 464\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 464\n",
      "       total no. variants observed              : 27315\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.43s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "This analysis is totally hard-core!\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 253342 sequences (63309110 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.989 CPU sec, 4.212 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 253342 reads in 31.448 CPU sec, 3.854 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R[samclip] Total SAM records 574536, removed 9650, allowed 1603, passed 564886\n",
      "[samclip] Header contained 3 lines\n",
      " @RG\\tID:ERR10479514\\tSM:ERR10479514 -t[samclip] Done.\n",
      " 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 10.488 sec; CPU: 67.025 sec\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 61\n",
      "          no. right trimmed                     : 393\n",
      "          no. left and right trimmed            : 18\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 472\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 472\n",
      "       total no. variants observed              : 27309\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.45s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "Run 'snp-sites -b -c -o phylo.aln core.full.aln' for IQTree, BEAST, RaxML\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 252588 sequences (63122680 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.892 CPU sec, 4.174 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 252588 reads in 28.847 CPU sec, 3.524 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479515\\tSM:ERR10479515 -t 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.457 sec; CPU: 63.969 sec\n",
      "[samclip] Total SAM records 573763, removed 9498, allowed 1571, passed 564265\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 57\n",
      "          no. right trimmed                     : 387\n",
      "          no. left and right trimmed            : 15\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 459\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 459\n",
      "       total no. variants observed              : 27299\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.44s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "5\tERR10479515\tsnp=29315\tdel=603\tins=762\thet=709\tunaligned=306177\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "The Snippy manual is at http://github.com/tseemann/snippy/blob/master/README.md\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 252052 sequences (62942802 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.593 CPU sec, 4.145 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 252052 reads in 28.349 CPU sec, 3.465 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479516\\tSM:ERR10479516 -t 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.745 sec; CPU: 63.403 sec\n",
      "[samclip] Total SAM records 573365, removed 9565, allowed 1648, passed 563800\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 55\n",
      "          no. right trimmed                     : 386\n",
      "          no. left and right trimmed            : 18\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 459\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 459\n",
      "       total no. variants observed              : 27283\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.47s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "5\tERR10479515\tsnp=29315\tdel=603\tins=762\thet=709\tunaligned=306177\n",
      "6\tERR10479516\tsnp=29293\tdel=622\tins=792\thet=775\tunaligned=305541\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "The Snippy manual is at http://github.com/tseemann/snippy/blob/master/README.md\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 252422 sequences (63033870 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 34.936 CPU sec, 4.317 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 252422 reads in 30.570 CPU sec, 3.726 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R[samclip] Total SAM records 573735, removed 9500, allowed 1661, passed 564235\n",
      "[samclip] Header contained 3 lines\n",
      " @RG\\tID:ERR10479517\\tSM:ERR10479517 -t[samclip] Done.\n",
      " 8 reference/ref.fa fake_reads.fq\n",
      "[main] Real time: 9.956 sec; CPU: 66.890 sec\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 56\n",
      "          no. right trimmed                     : 386\n",
      "          no. left and right trimmed            : 18\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 460\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 460\n",
      "       total no. variants observed              : 27305\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.44s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "5\tERR10479515\tsnp=29315\tdel=603\tins=762\thet=709\tunaligned=306177\n",
      "6\tERR10479516\tsnp=29293\tdel=622\tins=792\thet=775\tunaligned=305541\n",
      "7\tERR10479517\tsnp=29336\tdel=637\tins=792\thet=726\tunaligned=305577\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "Run 'snp-sites -b -c -o phylo.aln core.full.aln' for IQTree, BEAST, RaxML\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 265630 sequences (66341968 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.658 CPU sec, 4.131 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 265630 reads in 31.763 CPU sec, 3.880 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479518\\tSM:ERR10479518 -t 8 reference/ref.fa fake_reads.fq[samclip] Total SAM records 586987, removed 9619, allowed 1635, passed 577368\n",
      "\n",
      "[main] Real time: 10.016 sec; CPU: 66.788 sec\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 57\n",
      "          no. right trimmed                     : 384\n",
      "          no. left and right trimmed            : 18\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 459\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 459\n",
      "       total no. variants observed              : 27292\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.44s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "5\tERR10479515\tsnp=29315\tdel=603\tins=762\thet=709\tunaligned=306177\n",
      "6\tERR10479516\tsnp=29293\tdel=622\tins=792\thet=775\tunaligned=305541\n",
      "7\tERR10479517\tsnp=29336\tdel=637\tins=792\thet=726\tunaligned=305577\n",
      "8\tERR10479518\tsnp=29321\tdel=607\tins=777\thet=742\tunaligned=305529\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "This analysis is totally hard-core!\n",
      "Done.\n",
      "[M::bwa_idx_load_from_disk] read 0 ALT contigs\n",
      "[samclip] samclip 0.4.0 by Torsten Seemann (@torstenseemann)\n",
      "[samclip] Loading: reference/ref.fa.fai\n",
      "[samclip] Found 1 sequences in reference/ref.fa.fai\n",
      "[M::process] read 320000 sequences (80000000 bp)...\n",
      "[M::process] read 252874 sequences (63131650 bp)...\n",
      "[M::mem_process_seqs] Processed 320000 reads in 33.673 CPU sec, 4.153 real sec\n",
      "[samclip] Processed 100000 records...\n",
      "[samclip] Processed 200000 records...\n",
      "[samclip] Processed 300000 records...\n",
      "[M::mem_process_seqs] Processed 252874 reads in 29.220 CPU sec, 3.567 real sec\n",
      "[samclip] Processed 400000 records...\n",
      "[samclip] Processed 500000 records...\n",
      "[main] Version: 0.7.17-r1188\n",
      "[main] CMD: bwa mem -Y -M -R @RG\\tID:ERR10479519\\tSM:ERR10479519 -t 8 reference/ref.fa fake_reads.fq[samclip] Total SAM records 574230, removed 9558, allowed 1557, passed 564672\n",
      "\n",
      "[main] Real time: 9.765 sec; CPU: 64.086 sec\n",
      "[samclip] Header contained 3 lines\n",
      "[samclip] Done.\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 3 in-memory blocks...\n",
      "normalize v0.5\n",
      "\n",
      "options:     input VCF file                                  -\n",
      "         [o] output VCF file                                 -\n",
      "         [w] sorting window size                             10000\n",
      "         [n] no fail on reference inconsistency for non SNPs false\n",
      "         [q] quiet                                           false\n",
      "         [d] debug                                           false\n",
      "         [r] reference FASTA file                            reference/ref.fa\n",
      "\n",
      "\n",
      "stats: biallelic\n",
      "          no. left trimmed                      : 56\n",
      "          no. right trimmed                     : 388\n",
      "          no. left and right trimmed            : 17\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. biallelic normalized           : 461\n",
      "\n",
      "       multiallelic\n",
      "          no. left trimmed                      : 0\n",
      "          no. right trimmed                     : 0\n",
      "          no. left and right trimmed            : 0\n",
      "          no. right trimmed and left aligned    : 0\n",
      "          no. left aligned                      : 0\n",
      "\n",
      "       total no. multiallelic normalized        : 0\n",
      "\n",
      "       total no. variants normalized            : 461\n",
      "       total no. variants observed              : 27292\n",
      "       total no. reference observed             : 0\n",
      "\n",
      "Time elapsed: 0.43s\n",
      "\n",
      "This is snippy-core 4.6.0\n",
      "Obtained from http://github.com/tseemann/snippy\n",
      "Enabling bundled tools for linux\n",
      "Found any2fasta - /usr/local/bin/any2fasta\n",
      "Found samtools - /usr/local/bin/samtools\n",
      "Found minimap2 - /usr/local/bin/minimap2\n",
      "Found bedtools - /usr/local/bin/bedtools\n",
      "Found snp-sites - /usr/local/bin/snp-sites\n",
      "Saving reference FASTA: core.ref.fa\n",
      "This is any2fasta 0.4.2\n",
      "Opening '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Detected FASTA format\n",
      "Read 78307 lines from '/home/jupyter-mdprieto/tools/reference_data/GCF_000006765.1_ASM676v1_genomic.fna'\n",
      "Wrote 1 sequences from FASTA file.\n",
      "Processed 1 files.\n",
      "Done.\n",
      "Loaded 1 sequences totalling 6264404 bp.\n",
      "Will mask 0 regions totalling 0 bp ~ 0.00%\n",
      "0\tERR10479510\tsnp=29320\tdel=604\tins=821\thet=716\tunaligned=305610\n",
      "1\tERR10479511\tsnp=29300\tdel=642\tins=792\thet=726\tunaligned=309538\n",
      "2\tERR10479512\tsnp=29314\tdel=626\tins=792\thet=711\tunaligned=305996\n",
      "3\tERR10479513\tsnp=29345\tdel=627\tins=791\thet=709\tunaligned=305480\n",
      "4\tERR10479514\tsnp=29328\tdel=668\tins=794\thet=694\tunaligned=306415\n",
      "5\tERR10479515\tsnp=29315\tdel=603\tins=762\thet=709\tunaligned=306177\n",
      "6\tERR10479516\tsnp=29293\tdel=622\tins=792\thet=775\tunaligned=305541\n",
      "7\tERR10479517\tsnp=29336\tdel=637\tins=792\thet=726\tunaligned=305577\n",
      "8\tERR10479518\tsnp=29321\tdel=607\tins=777\thet=742\tunaligned=305529\n",
      "9\tERR10479519\tsnp=29311\tdel=664\tins=774\thet=725\tunaligned=306694\n",
      "Opening: core.tab\n",
      "Opening: core.vcf\n",
      "Processing contig: NC_002516.2\n",
      "Generating core.full.aln\n",
      "Creating TSV file: core.txt\n",
      "Running: snp-sites -c -o core.aln core.full.aln\n",
      "This analysis is totally hard-core!\n",
      "Done.\n"
     ]
    }
   ],
   "source": [
    "for contig in $(ls $HOME/tutorials/contigs/*contigs.fa)\n",
    "do\n",
    "    # define isolate name\n",
    "    meta=$(basename \"$contig\" '_contigs.fa')\n",
    "\n",
    "    # run snippy for each isolate\n",
    "    singularity exec \"$SNIPPY_IMG\" snippy \\\n",
    "    --outdir \"$HOME/tutorials/results_snippy/$meta\"  `# save in a subdirectory`\\\n",
    "    --ctgs \"$contig\" \\\n",
    "    --ref $REF_GENOME \\\n",
    "    --cpus 8 \\\n",
    "    --force \\\n",
    "    --cleanup \\\n",
    "    --quiet\n",
    "\n",
    "    # produce core genome analysis in results_snippy folder\n",
    "    cd $HOME/tutorials/results_snippy/\n",
    "    singularity exec \"$SNIPPY_IMG\" snippy-core --ref $REF_GENOME \\\n",
    "    /home/jupyter-mdprieto/tutorials/results_snippy/ERR*\n",
    "done"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "d10ecbf5",
   "metadata": {},
   "source": [
    "Once we have the data from the snippy call, we run a few additional steps to clean our results and to have them ready for phylogenetic analysis. \n",
    "\n",
    "1. We run **snippy-core** to make a summary of SNV found in genes that are common to all isolates analyzed\n",
    "2. We run another tool, **Gubbins**, that eliminates features that are known to be transmitted through horizontal gene transfer. As the variability in these may come from other bacteria in the environment, they are not useful for phylogenetic analysis. \n",
    "3. Once we have the cleaned alignment, we use **snp-sites** to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94d37a32",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cd $HOME/tutorials/results_snippy\n",
    "\n",
    "# clean file\n",
    "singularity exec $SNIPPY_IMG snippy-clean_full_aln core.full.aln > clean.full.aln\n",
    "\n",
    "# gubbins remove horizontal gene transfer \n",
    "singularity exec $GUBBINS_IMG run_gubbins.py -p gubbins clean.full.aln\n",
    "\n",
    "# SNP-sites\n",
    "singularity exec $SNIPPY_IMG snp-sites -c gubbins.filtered_polymorphic_sites.fasta > clean.core.aln\n",
    "\n",
    "# produce tree\n",
    "$HOME/tools/FastTree -gtr -nt clean.core.aln > clean.core.tree"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "id": "866f3030",
   "metadata": {},
   "source": [
    "Finally, we can visualize the tree we just produce by pasting the contents of the file `clean.core.tree` in a visualizer tool such as [phyl.io](https://phylo.io/), or [iTOL](https://itol.embl.de/upload.cgi).\n",
    "\n",
    "The data we used may not be the best representation of differences in a phylogenetic tree as it comes from a single center outbreak and the samples were collected close to each other. Another issue may be that the reference genome employed was not adequate and thus, it is not showing significant differences. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "d84a5b12",
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "cd $CONTIGS_DIR\n",
    "readlink -f *contigs.fa > path.txt\n",
    "cat path.txt | sed -s \"s/_contigs.fa//\" | sed -s 's/contigs/results_snippy/' > filenames.txt\n",
    "\n",
    "\n",
    "# join them and remove temp files\n",
    "paste filenames.txt path.txt > ~/scripts/snippy-input.tab\n",
    "rm filenames.txt path.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
